src/clipping/weileratherton.jl

```diff
@@ -0,0 +1,417 @@
+
+function collectoutsiderings(ring, polyrings)
+  newpolyrings = Ring[]
+  for k in eachindex(polyrings)
+    if !all(v ∈ PolyArea(ring) for v in vertices(polyrings[k]))
+      push!(newpolyrings, polyrings[k])
+    end
+  end
+  if any(v ∈ PolyArea(polyrings[1]) for v in vertices(ring))
+    push!(newpolyrings, ring)
+  end
+  newpolyrings
+end
+
+function perhapsaddnonintersected!(ring, polyrings, intersected)
+  # Handle all the non-intersected rings. Add them if they are contained in the clipping ring.
+  newpolyrings = [ring]
+  for k in eachindex(intersected)
+    if !intersected[k]
+      ccw = orientation(ring) == CCW
+
+      # Discard if the processed ring is an outer ring but inside another inner ring.
+      if ccw && vertices(ring)[1] ∈ PolyArea(polyrings[k]) && k > 1
+        return nothing
+      end
+
+      vs = vertices(polyrings[k])
+      ins = count(v ∈ PolyArea(ring) for v in vs)
+      outs = length(vs) - ins
+      if ccw == (ins > outs)
+        if !ccw
+          # An outer ring should be first in the list.
+          pushfirst!(newpolyrings, polyrings[k])
+        else
+          # Inner rings should follow an outer in the list.
+          push!(newpolyrings, polyrings[k])
+        end
+        intersected[k] = true
+      end
+    end
+  end
+  newpolyrings
+end
+
+# Inserts the intersection in the ring.
+function insertintersection!(head::RingVertex, intersection::RingVertex, side::Symbol)
+  tail = head.left
+  vertex = head
+
+  new = measure(Segment(head.point, intersection.point))
+  while true
+    os = isnormal(vertex) ? :right : side
+    current = measure(Segment(head.point, getfield(vertex, os).point))
+    if (new < current) || (getfield(vertex, os) == tail)
+      next = getfield(vertex, os)
+      if !(new ≈ measure(Segment(head.point, next.point)))
+        setfield!(intersection, side, next)
+        setfield!(vertex, os, intersection)
+      end
+      break
+    end
+    vertex = getfield(vertex, os)
+  end
+end
+
+# Inserts the intersection into both the clipping and the clipped rings.
+function insertintersections!(vertex::Tuple, clipping, clipped, entering)
+  (vtype, point) = vertex
+  if !isnothing(vtype)
+    intersection = RingVertex{vtype}(point)
+    insertintersection!(clipping.left, intersection, :left)
+    insertintersection!(clipped.left, intersection, :right)
+
+    if vtype == Entering
+      push!(entering, intersection)
+    end
+    return true
+  end
+  false
+end
+
+insertintersections!(vertices::Array, clipping, clipped, entering) =
+  any(insertintersections!.(vertices, Ref(clipping), Ref(clipped), Ref(entering)))
+
+# Traversing and selecting following vertices for collecting of the rings after clipping depends
+# on the vertex type. Helper nextvertex follows the correct path.
+nextvertex(v::RingVertex{Normal}) = v.right
+nextvertex(v::RingVertex{Entering}) = v.right
+nextvertex(v::RingVertex{Exiting}) = v.left
```

Comment:
This is part of the RingVertex api, please move it to after the RingVertex definition. We will take a look later if these new VertexType and RingVertex are really necessary or if we can rely on simple integers/enums.